### PR TITLE
implement slice_string/3

### PR DESCRIPTION
**Pull request by DeemoONeill**
add implementation for slicing strings. 

implements a slice_string/2 and a slice_string/3, the former just takes an offset and the latter takes an offset and length

references https://github.com/elixir-explorer/explorer/issues/663

### CODE REVIEW

File: lib/explorer/backend/lazy_series.ex [hunk 1]
```diff
@@ -116,6 +116,7 @@ defmodule Explorer.Backend.LazySeries do
     trim: 2,
     upcase: 1,
     downcase: 1,
+    slice_string: 3,
```
Comment (DeemoONeill) [inline lib/explorer/backend/lazy_series.ex:119]: i'm unsure what to call this as slice for arrays

**Comment by DeemoONeill** [comment]
this also took me longer than i would like to admit to realise i had to add the new functions to lib.rs. I must have run the tests 50 times getting `nif not found` before i realised 🤦‍♂️

File: lib/explorer/series.ex [hunk 2]
```diff
@@ -4355,6 +4355,61 @@ defmodule Explorer.Series do
 
   def trim_trailing(%Series{dtype: dtype}), do: dtype_error("trim_trailing/1", dtype, [:string])
 
+  @doc """
+  Returns a string sliced from the offset to the end of the string, supporting
+  negative indexing
```
Comment (josevalim) [inline lib/explorer/series.ex:4360]: I assume this works on Unicode, so can we add tests for something like: こんにちは世界?

File: lib/explorer/backend/lazy_series.ex (hunk 1 above)
Comment (josevalim) [reply lib/explorer/backend/lazy_series.ex:119]: Yeah, I am not sure what to call it either. And I assume that, if we want to slice lists later, we don't need a list_slice, we would simply overload this function, so we shouldn't have a `_string` suffix. It could be `slice_value` or `slice_contents` or something, but I am not sure.

Comment (DeemoONeill) [reply lib/explorer/backend/lazy_series.ex:119]: Maybe I can overload the existing slice to work when dtype is string. I'll check tomorrow

Comment (josevalim) [reply lib/explorer/backend/lazy_series.ex:119]: That won't work because you may still want to slice a string series both ways. :( by list, I meant when we have a series of lists in the future!

Comment (cigrainger) [reply lib/explorer/backend/lazy_series.ex:119]: I'd go with `substring`, e.g. https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/substr and https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substring. We don't necessary _have_ to overload it for lists, right?

Comment (DeemoONeill) [reply lib/explorer/backend/lazy_series.ex:119]: If we call it substring we have the same issue if/when similar functionality is implemented for lists though right?

In that instance would it be called `sublist`?

It's a slight shame its not namespaced. Then `Series.slice` would slice the series. `Series.String.slice` would slice strings etc.

I'm not sure... I'm happy to go any way on this

Comment (josevalim) [reply lib/explorer/backend/lazy_series.ex:119]: I actually think the namespacing here would not help because using the same function for doing very different things could be confusing. :( I think we have two options here:

1. `substring` as proposed by @cigrainger 

2. `string_part` (to mirror Elixir's `binary_part`) - however, `binary_part` is strict, `binary_part("elixir", 0, 13)` raises if "length" falls outside and it does not support a negative index, so I don't think it is suitable.

Well, that said, then :+1: for `substring` and we might add `sublist` and `subbinary` in the future if necessary. Can you please amend? Also please add tests for things like both start and length falling out of range. :)

File: lib/explorer/series.ex [hunk 3]
```diff
@@ -4355,6 +4355,61 @@ defmodule Explorer.Series do
...
+      #Explorer.Series<
+        Polars[3]
+        string ["arth", "ars", "eptun"]
+      >
+  """
+  @doc type: :string_wise
+  @spec slice_string(Series.t(), integer(), integer()) :: Series.t()
+  def slice_string(%Series{dtype: :string} = series, offset, length)
+      when is_integer(offset)
+      when is_integer(length),
```
Comment (josevalim) [inline lib/explorer/series.ex:4407]:
```suggestion
      when is_integer(length) and length >= 0,
```

File: lib/explorer/series.ex [hunk 4]
```diff
@@ -4355,6 +4355,61 @@ defmodule Explorer.Series do
...
+
+      iex> s = Explorer.Series.from_list(["earth", "mars", "neptune"])
+      iex> Explorer.Series.slice_string(s, 1, 5)
+      #Explorer.Series<
+        Polars[3]
+        string ["arth", "ars", "eptun"]
+      >
+  """
+  @doc type: :string_wise
+  @spec slice_string(Series.t(), integer(), integer()) :: Series.t()
```
Comment (josevalim) [inline lib/explorer/series.ex:4404]:
```suggestion
  @spec slice_string(Series.t(), integer(), non_neg_integer()) :: Series.t()
```

File: lib/explorer/backend/lazy_series.ex (hunk 1 above)
Comment (DeemoONeill) [reply lib/explorer/backend/lazy_series.ex:119]: I'll get on it after work. Thanks for the input both :)

File: lib/explorer/series.ex (hunk 2 above)
Comment (DeemoONeill) [reply lib/explorer/series.ex:4360]: added

**Comment by josevalim** [comment]
This is awesome @DeemoONeill, thank you for taking the time to improve Explorer! If you need further APIs in the future, don't hesitate to start a discussion! :green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Comment by DeemoONeill** [comment]
Thanks José, appreciate the feedback and help!
